File: 백준/Gold/1987. 알파벳/알파벳.py
import sys

dxs, dys = (1, -1, 0, 0), (0, 0, 1, -1)
n, m = map(int, sys.stdin.readline().rstrip().split())
g = [list(map(str, sys.stdin.readline().rstrip())) for _ in range(n)]
visited = 0
visited |= (1 << (ord(g[0][0]) - 65))
ans = 0


def dfs(x, y, cnt):
    global ans, visited
    ans = max(cnt, ans)
    for i in range(4):
        nx, ny = x + dxs[i], y + dys[i]
        if 0 <= nx < m and 0 <= ny < n and not (visited & (1 << (ord(g[ny][nx]) - 65))):
            visited |= (1 << (ord(g[ny][nx]) - 65))
            dfs(nx, ny, cnt + 1)
            visited &= ~(1 << (ord(g[ny][nx]) - 65))


dfs(0, 0, 1)
print(ans)

File: 백준/Gold/1987. 알파벳/알파벳.cc
#include <bits/stdc++.h>
using namespace std;
char g[21][21];
int ans = 0, n, m, dxs[4] = {1, -1, 0, 0}, dys[4] = {0, 0, 1, -1};
bool visited[26];
void dfs(int x, int y, int cnt){
    ans = max(ans, cnt);
    for (int i = 0; i < 4; i++){
        int nx = x + dxs[i], ny = y + dys[i];
        if (0 <= nx && nx < m && 0 <= ny && ny < n && !visited[(int)g[ny][nx] - 65]){
            visited[(int)g[ny][nx] - 65] = true;
            dfs(nx, ny, cnt + 1);
            visited[(int)g[ny][nx] - 65] = false;
        }
    }
}

int main(){
    cin >> n >> m;
    for(int i = 0; i < n; i++)
        for(int j = 0; j < m; j++)
            cin >> g[i][j];
    visited[(int)g[0][0] - 65] = true;
    dfs(0, 0, 1);
    cout << ans << '\n';
    return 0;
}